File: mhycli/bili_login.py
"""B站 TV 端二维码登录 — 复用 biliup credential.rs 逻辑

流程 (biliup get_qrcode + login_by_qrcode):
  1. POST /x/passport-tv-login/qrcode/auth_code → 获取 auth_code + 二维码 URL
  2. 用 qrcode_url 生成二维码, 手机 B 站 APP 扫码
  3. 轮询 POST /x/passport-tv-login/qrcode/poll:
     - code=0     → 登录成功, 返回 cookie_info (含 SESSDATA 等)
     - code=86039 → 尚未确认, 继续轮询
  登录成功后可保存 cookie 到 Config/bili_cookie.json (get_bili_cookie 读取格式)
"""
from __future__ import annotations

import hashlib
import json
import time
import uuid
from pathlib import Path

import requests

# BiliTV app_key/appsec (biliup AppKeyStore::BiliTV)
BILI_TV_APPKEY = "4409e2ce8ffd12b8"
BILI_TV_APPSEC = "59b43e04ad6965f34319062b478f83dd"

# TV 端登录 API
API_AUTH_CODE = "https://passport.bilibili.com/x/passport-tv-login/qrcode/auth_code"
API_POLL = "https://passport.bilibili.com/x/passport-tv-login/qrcode/poll"

# BiliApp UA (biliup login_by_web_qrcode 使用; 缺 UA 会返回 412)
BILI_UA = ("Mozilla/5.0 (X11; Linux x86_64; rv:38.0) Gecko/20100101 "
           "Firefox/38.0 Iceweasel/38.2.1 BiliApp")

# cookie 保存路径 (与 get_bili_cookie 一致)
COOKIE_FILE = Path(__file__).resolve().parent.parent / "Config" / "bili_cookie.json"


def _sign(urlencoded: str, app_sec: str = BILI_TV_APPSEC) -> str:
    """MD5(urlencoded_params + appsec) — 对应 biliup sign()"""
    return hashlib.md5((urlencoded + app_sec).encode("utf-8")).hexdigest()


def _build_auth_code_form() -> dict:
    """构造 auth_code 请求 form (appkey + local_id + ts + sign)"""
    from urllib.parse import urlencode

    form = {
        "appkey": BILI_TV_APPKEY,
        "local_id": "0",
        "ts": str(int(time.time())),
    }
    form["sign"] = _sign(urlencode(form))
    return form


def _tv_headers() -> dict:
    """TV 端请求头 (BiliApp UA 必须, 否则 412)"""
    return {"User-Agent": BILI_UA}


def get_qrcode(session: requests.Session | None = None) -> tuple[str, str]:
    """获取登录二维码, 返回 (qrcode_url, auth_code)

    对应 biliup get_qrcode
    """
    s = session or requests.Session()
    resp = s.post(API_AUTH_CODE, data=_build_auth_code_form(),
                  headers=_tv_headers(), timeout=15)
    resp.raise_for_status()
    j = resp.json()
    if j.get("code") != 0:
        raise RuntimeError(f"获取二维码失败: {j.get('code')} {j.get('message')}")
    data = j.get("data", {})
    qrcode_url = data.get("url") or data.get("qrcode_url") or ""
    auth_code = data.get("auth_code") or ""
    if not qrcode_url or not auth_code:
        raise RuntimeError(f"二维码数据不完整: {data}")
    return qrcode_url, auth_code


def poll_login(session: requests.Session | None, auth_code: str,
               on_status=None, timeout: float = 180.0) -> dict:
    """轮询扫码状态, 登录成功返回完整响应 (含 data.cookie_info)

    对应 biliup login_by_qrcode
    on_status(code): 状态回调 (86039=等待确认, -4=未扫, -5=已扫, 0=成功)
    """
    from urllib.parse import urlencode

    s = session or requests.Session()
    deadline = time.time() + timeout
    while time.time() < deadline:
        form = {
            "appkey": BILI_TV_APPKEY,
            "auth_code": auth_code,
            "local_id": "0",
            "ts": str(int(time.time())),
        }
        form["sign"] = _sign(urlencode(form))
        try:
            resp = s.post(API_POLL, data=form, headers=_tv_headers(), timeout=15)
            resp.raise_for_status()
            j = resp.json()
        except (requests.RequestException, ValueError) as e:
            if on_status:
                on_status("network_error")
            time.sleep(2)
            continue

        code = j.get("code")
        if on_status:
            on_status(code)

        if code == 0 and j.get("data", {}).get("cookie_info"):
            return j  # 登录成功
        if code == -4:
            continue  # 二维码未扫描
        time.sleep(1.5)
    raise TimeoutError("扫码登录超时")


def extract_cookies(login_response: dict) -> list[dict]:
    """从登录响应提取 cookie 列表 [{name, value}]"""
    cookie_info = (login_response.get("data") or {}).get("cookie_info") or {}
    return list(cookie_info.get("cookies") or [])


def save_cookies_to_file(cookies: list[dict], path: Path | None = None):
    """保存 cookie 到 biliup 格式文件 (get_bili_cookie 读取格式)"""
    target = path or COOKIE_FILE
    target.parent.mkdir(parents=True, exist_ok=True)
    data = {"cookie_info": {"cookies": cookies}}
    target.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")


def cookie_summary(cookies: list[dict]) -> str:
    """返回 cookie 概要 (列出关键字段名)"""
    names = [c.get("name") for c in cookies if c.get("name")]
    return ", ".join(names) if names else "(空)"

File: src/bili_login.rs
//! B站 TV 端二维码登录（保存拉流 cookie），对应 Python `bili_login.py` / biliup。
use std::path::PathBuf;

use anyhow::{Result, bail};
use reqwest::header::{CONTENT_TYPE, USER_AGENT};
use serde_json::Value;

use crate::crypto::{md5_hex, now_secs};

const APPKEY: &str = "4409e2ce8ffd12b8";
const APPSEC: &str = "59b43e04ad6965f34319062b478f83dd";
const UA: &str =
    "Mozilla/5.0 (X11; Linux x86_64; rv:38.0) Gecko/20100101 Firefox/38.0 Iceweasel/38.2.1 BiliApp";
const API_AUTH_CODE: &str = "https://passport.bilibili.com/x/passport-tv-login/qrcode/auth_code";
const API_POLL: &str = "https://passport.bilibili.com/x/passport-tv-login/qrcode/poll";

fn form_encode(pairs: &[(&str, &str)]) -> String {
    let mut s = url::form_urlencoded::Serializer::new(String::new());
    for (k, v) in pairs {
        s.append_pair(k, v);
    }
    s.finish()
}

/// MD5(urlencoded_params + appsec)
fn sign(form: &str) -> String {
    md5_hex(&format!("{form}{APPSEC}"))
}

/// 获取登录二维码，返回 (qrcode_url, auth_code)
pub async fn get_qrcode(http: &reqwest::Client) -> Result<(String, String)> {
    let ts = now_secs().to_string();
    let base = form_encode(&[("appkey", APPKEY), ("local_id", "0"), ("ts", &ts)]);
    let body = format!("{base}&sign={}", sign(&base));

    let resp: Value = http
        .post(API_AUTH_CODE)
        .header(USER_AGENT, UA)
        .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
        .body(body)
        .send()
        .await?
        .json()
        .await?;

    if resp.get("code").and_then(Value::as_i64) != Some(0) {
        bail!(
            "获取二维码失败: {} {}",
            resp.get("code").and_then(Value::as_i64).unwrap_or(-1),
            resp.get("message").and_then(Value::as_str).unwrap_or("")
        );
    }
    let data = resp.get("data").cloned().unwrap_or_default();
    let url = data
        .get("url")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string();
    let auth_code = data
        .get("auth_code")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string();
    if url.is_empty() || auth_code.is_empty() {
        bail!("二维码数据不完整: {data}");
    }
    Ok((url, auth_code))
}

/// 轮询扫码状态，成功返回完整响应（含 data.cookie_info）
pub async fn poll_login(
    http: &reqwest::Client,
    auth_code: &str,
    timeout: std::time::Duration,
    log: &(dyn Fn(String) + Send + Sync),
) -> Result<Value> {
    let deadline = std::time::Instant::now() + timeout;
    let mut last_code: i64 = -999;
    while std::time::Instant::now() < deadline {
        let ts = now_secs().to_string();
        let base = form_encode(&[
            ("appkey", APPKEY),
            ("auth_code", auth_code),
            ("local_id", "0"),
            ("ts", &ts),
        ]);
        let body = format!("{base}&sign={}", sign(&base));
        let resp = http
            .post(API_POLL)
            .header(USER_AGENT, UA)
            .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
            .body(body)
            .send()
            .await;
        let j: Value = match resp {
            Ok(r) => match r.json().await {
                Ok(v) => v,
                Err(_) => {
                    tokio::time::sleep(std::time::Duration::from_secs(2)).await;
                    continue;
                }
            },
            Err(_) => {
                tokio::time::sleep(std::time::Duration::from_secs(2)).await;
                continue;
            }
        };
        let code = j.get("code").and_then(Value::as_i64).unwrap_or(-1);
        if code != last_code {
            let msg = match code {
                -4 => "二维码未扫描",
                -5 => "已扫描，等待确认",
                86039 => "已扫描，等待确认",
                86038 => "二维码已失效",
                0 => "登录成功",
                _ => "",
            };
            if !msg.is_empty() {
                log(format!("  扫码状态: {msg} (code={code})"));
            }
            last_code = code;
        }
        if code == 0 && j.get("data").and_then(|d| d.get("cookie_info")).is_some() {
            return Ok(j);
        }
        if code == 86038 {
            bail!("二维码已失效，请重试");
        }
        tokio::time::sleep(std::time::Duration::from_millis(1500)).await;
    }
    bail!("扫码登录超时")
}

/// 从登录响应提取 cookie 列表 [{name,value}]
pub fn extract_cookies(login_response: &Value) -> Vec<Value> {
    login_response
        .get("data")
        .and_then(|d| d.get("cookie_info"))
        .and_then(|c| c.get("cookies"))
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
}

fn cookie_path() -> PathBuf {
    if let Ok(p) = std::env::var("MHYSCAN_CONFIG") {
        if let Some(dir) = PathBuf::from(p).parent() {
            return dir.join("bili_cookie.json");
        }
    }
    PathBuf::from("Config/bili_cookie.json")
}

/// 保存 cookie（biliup 格式，供拉流复用）
pub fn save_cookies(cookies: &[Value]) -> Result<PathBuf> {
    let path = cookie_path();
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    let data = serde_json::json!({ "cookie_info": { "cookies": cookies } });
    std::fs::write(&path, serde_json::to_string_pretty(&data)?)?;
    Ok(path)
}

pub fn cookie_summary(cookies: &[Value]) -> String {
    let names: Vec<&str> = cookies
        .iter()
        .filter_map(|c| c.get("name").and_then(Value::as_str))
        .collect();
    if names.is_empty() {
        "(空)".into()
    } else {
        names.join(", ")
    }
}
